dagshub/data_engine/model/datasource.py

```diff
@@ -1064,3 +1195,119 @@ def wrap_bytes(val: bytes) -> str:
 
     def get_metadata_entries(self):
         return self._metadata_entries
+
+
+@dataclass
+class DatasourceQuery(DataClassJsonMixin):
+    as_of: Optional[int] = field(default=None, metadata=config(exclude=exclude_if_none, letter_case=LetterCase.CAMEL))
+    select: Optional[List] = field(default=None, metadata=config(exclude=exclude_if_none))
+    filter: "QueryFilterTree" = field(
+        default=QueryFilterTree(),
+        metadata=config(field_name="query", encoder=QueryFilterTree.serialize, decoder=QueryFilterTree.deserialize),
+    )
+
+    def __deepcopy__(self, memodict={}):
+        other = DatasourceQuery(
+            as_of=self.as_of,
+            filter=self.filter.__deepcopy__(),
+        )
+        if self.select is not None:
+            other.select = self.select.copy()
+        return other
```

Comment:
Can't you use just json serder for this? Or you don't want to for efficiency?

Reply:
This is running every time you add a new filter/change the datasource in any way, so yea, I try not to waste time on serializing here